### ISSUE TITLE
Example failing on deserializing option types

### ISSUE THREAD

**Issue by wiredsister** [usix79/Avro.FSharp#3]
Hi there, 

First of all, I want to thank you so much for your amazing Avro example for binary serialization. 

I have a simple domain app which is a Burger restaurant that takes orders and sends receipts to Kafka. The domain that gets serialized to an Avro type is:

```fsharp 

module Receipt =

  type T = {
    Customer: Customer.T
    OrderID: Guid
    OrderNumber: int64
    SalesTax: decimal
    Items: Item.T list
    Time: DateTime
    Total: decimal
  }

```

Originally, I made the "sales tax"  a `decimal option` type like: 

```fsharp 

module Receipt =

  type T = {
    Customer: Customer.T
    OrderID: Guid
    OrderNumber: int64
    SalesTax: decimal option
    Items: Item.T list
    Time: DateTime
    Total: decimal
  }

```

Unfortunately, when it came to using the example that you provided with that binary Avro deserialization I got null generic system exceptions that were difficult to debug in the consumer. Here's an example of my logs:

```
example-consumer  | |SCHEMA|EXN:System.Exception: Unknown type: null
example-consumer  |    at Microsoft.FSharp.Core.PrintfModule.PrintFormatToStringThenFail@1448.Invoke(String message)
example-consumer  |    at Avro.FSharp.SchemaModule.parse@316-3.GenerateNext(IEnumerable`1& next)
example-consumer  |    at Microsoft.FSharp.Core.CompilerServices.GeneratedSequenceBase`1.MoveNextImpl() in D:\a\_work\1\s\src\FSharp.Core\seqcore.fs:line 488
example-consumer  |    at Microsoft.FSharp.Core.CompilerServices.GeneratedSequenceBase`1.System.Collections.IEnumerator.MoveNext() in D:\a\_work\1\s\src\FSharp.Core\seqcore.fs:line 530
example-consumer  |    at Microsoft.FSharp.Collections.SeqModule.ToArray[T](IEnumerable`1 source) in D:\a\_work\1\s\src\FSharp.Core\seq.fs:line 1027
example-consumer  |    at Microsoft.FSharp.Collections.ArrayModule.OfSeq[T](IEnumerable`1 source) in D:\a\_work\1\s\src\FSharp.Core\array.fs:line 1543
example-consumer  |    at Avro.FSharp.SchemaModule.parse@252-1(Dictionary`2 cache, String ns, JsonElement el)
example-consumer  |    at Avro.FSharp.SchemaModule.parse@252-1(Dictionary`2 cache, String ns, JsonElement el)
example-consumer  |    at Example.KafkaSerde.Message.Deserializer.Confluent.Kafka.IDeserializer<Example.Receipt.T>.Deserialize(ReadOnlySpan`1 data, Boolean isNull, SerializationContext context) in /lib/lib/Data.fs:line 128|
example-consumer  | |CONSUMER|ERROR:Local: Value deserialization error|
example-consumer  | |SCHEMA|EXN:System.Exception: Unknown type: null
example-consumer  |    at Microsoft.FSharp.Core.PrintfModule.PrintFormatToStringThenFail@1448.Invoke(String message)
example-consumer  |    at Avro.FSharp.SchemaModule.parse@316-3.GenerateNext(IEnumerable`1& next)
example-consumer  |    at Microsoft.FSharp.Core.CompilerServices.GeneratedSequenceBase`1.MoveNextImpl() in D:\a\_work\1\s\src\FSharp.Core\seqcore.fs:line 488
example-consumer  |    at Microsoft.FSharp.Core.CompilerServices.GeneratedSequenceBase`1.System.Collections.IEnumerator.MoveNext() in D:\a\_work\1\s\src\FSharp.Core\seqcore.fs:line 530
example-consumer  |    at Microsoft.FSharp.Collections.SeqModule.ToArray[T](IEnumerable`1 source) in D:\a\_work\1\s\src\FSharp.Core\seq.fs:line 1027
example-consumer  |    at Microsoft.FSharp.Collections.ArrayModule.OfSeq[T](IEnumerable`1 source) in D:\a\_work\1\s\src\FSharp.Core\array.fs:line 1543
example-consumer  |    at Avro.FSharp.SchemaModule.parse@252-1(Dictionary`2 cache, String ns, JsonElement el)
example-consumer  |    at Avro.FSharp.SchemaModule.parse@252-1(Dictionary`2 cache, String ns, JsonElement el)
example-consumer  |    at Example.KafkaSerde.Message.Deserializer.Confluent.Kafka.IDeserializer<Example.Receipt.T>.Deserialize(ReadOnlySpan`1 data, Boolean isNull, SerializationContext context) in /lib/lib/Data.fs:line 128|
example-consumer  | |CONSUMER|ERROR:Local: Value deserialization error|
example-consumer  | |SCHEMA|EXN:System.Exception: Unknown type: null
example-consumer  |    at Microsoft.FSharp.Core.PrintfModule.PrintFormatToStringThenFail@1448.Invoke(String message)
example-consumer  |    at Avro.FSharp.SchemaModule.parse@316-3.GenerateNext(IEnumerable`1& next)
example-consumer  |    at Microsoft.FSharp.Core.CompilerServices.GeneratedSequenceBase`1.MoveNextImpl() in D:\a\_work\1\s\src\FSharp.Core\seqcore.fs:line 488
example-consumer  |    at Microsoft.FSharp.Core.CompilerServices.GeneratedSequenceBase`1.System.Collections.IEnumerator.MoveNext() in D:\a\_work\1\s\src\FSharp.Core\seqcore.fs:line 530
example-consumer  |    at Microsoft.FSharp.Collections.SeqModule.ToArray[T](IEnumerable`1 source) in D:\a\_work\1\s\src\FSharp.Core\seq.fs:line 1027
example-consumer  |    at Microsoft.FSharp.Collections.ArrayModule.OfSeq[T](IEnumerable`1 source) in D:\a\_work\1\s\src\FSharp.Core\array.fs:line 1543
example-consumer  |    at Avro.FSharp.SchemaModule.parse@252-1(Dictionary`2 cache, String ns, JsonElement el)
example-consumer  |    at Avro.FSharp.SchemaModule.parse@252-1(Dictionary`2 cache, String ns, JsonElement el)
example-consumer  |    at Example.KafkaSerde.Message.Deserializer.Confluent.Kafka.IDeserializer<Example.Receipt.T>.Deserialize(ReadOnlySpan`1 data, Boolean isNull, SerializationContext context) in /lib/lib/Data.fs:line 128|
example-consumer  | |CONSUMER|ERROR:Local: Value deserialization error|
example-consumer  | |SCHEMA|EXN:System.Exception: Unknown type: null
example-consumer  |    at Microsoft.FSharp.Core.PrintfModule.PrintFormatToStringThenFail@1448.Invoke(String message)
example-consumer  |    at Avro.FSharp.SchemaModule.parse@316-3.GenerateNext(IEnumerable`1& next)
example-consumer  |    at Microsoft.FSharp.Core.CompilerServices.GeneratedSequenceBase`1.MoveNextImpl() in D:\a\_work\1\s\src\FSharp.Core\seqcore.fs:line 488
example-consumer  |    at Microsoft.FSharp.Core.CompilerServices.GeneratedSequenceBase`1.System.Collections.IEnumerator.MoveNext() in D:\a\_work\1\s\src\FSharp.Core\seqcore.fs:line 530
example-consumer  |    at Microsoft.FSharp.Collections.SeqModule.ToArray[T](IEnumerable`1 source) in D:\a\_work\1\s\src\FSharp.Core\seq.fs:line 1027
example-consumer  |    at Microsoft.FSharp.Collections.ArrayModule.OfSeq[T](IEnumerable`1 source) in D:\a\_work\1\s\src\FSharp.Core\array.fs:line 1543
example-consumer  |    at Avro.FSharp.SchemaModule.parse@252-1(Dictionary`2 cache, String ns, JsonElement el)
example-consumer  |    at Avro.FSharp.SchemaModule.parse@252-1(Dictionary`2 cache, String ns, JsonElement el)
example-consumer  |    at Example.KafkaSerde.Message.Deserializer.Confluent.Kafka.IDeserializer<Example.Receipt.T>.Deserialize(ReadOnlySpan`1 data, Boolean isNull, SerializationContext context) in /lib/lib/Data.fs:line 128|
example-consumer  | |CONSUMER|ERROR:Local: Value deserialization error|
example-consumer  | |SCHEMA|EXN:System.Exception: Unknown type: null
example-consumer  |    at Microsoft.FSharp.Core.PrintfModule.PrintFormatToStringThenFail@1448.Invoke(String message)
example-consumer  |    at Avro.FSharp.SchemaModule.parse@316-3.GenerateNext(IEnumerable`1& next)
example-consumer  |    at Microsoft.FSharp.Core.CompilerServices.GeneratedSequenceBase`1.MoveNextImpl() in D:\a\_work\1\s\src\FSharp.Core\seqcore.fs:line 488
example-consumer  |    at Microsoft.FSharp.Core.CompilerServices.GeneratedSequenceBase`1.System.Collections.IEnumerator.MoveNext() in D:\a\_work\1\s\src\FSharp.Core\seqcore.fs:line 530
example-consumer  |    at Microsoft.FSharp.Collections.SeqModule.ToArray[T](IEnumerable`1 source) in D:\a\_work\1\s\src\FSharp.Core\seq.fs:line 1027
example-consumer  |    at Microsoft.FSharp.Collections.ArrayModule.OfSeq[T](IEnumerable`1 source) in D:\a\_work\1\s\src\FSharp.Core\array.fs:line 1543
example-consumer  |    at Avro.FSharp.SchemaModule.parse@252-1(Dictionary`2 cache, String ns, JsonElement el)
example-consumer  |    at Avro.FSharp.SchemaModule.parse@252-1(Dictionary`2 cache, String ns, JsonElement el)
example-consumer  |    at Example.KafkaSerde.Message.Deserializer.Confluent.Kafka.IDeserializer<Example.Receipt.T>.Deserialize(ReadOnlySpan`1 data, Boolean isNull, SerializationContext context) in /lib/lib/Data.fs:line 128|
example-consumer  | |CONSUMER|ERROR:Local: Value deserialization error|
example-consumer  | |SCHEMA|EXN:System.Exception: Unknown type: null
example-consumer  |    at Microsoft.FSharp.Core.PrintfModule.PrintFormatToStringThenFail@1448.Invoke(String message)
example-consumer  |    at Avro.FSharp.SchemaModule.parse@316-3.GenerateNext(IEnumerable`1& next)
example-consumer  |    at Microsoft.FSharp.Core.CompilerServices.GeneratedSequenceBase`1.MoveNextImpl() in D:\a\_work\1\s\src\FSharp.Core\seqcore.fs:line 488
example-consumer  |    at Microsoft.FSharp.Core.CompilerServices.GeneratedSequenceBase`1.System.Collections.IEnumerator.MoveNext() in D:\a\_work\1\s\src\FSharp.Core\seqcore.fs:line 530
example-consumer  |    at Microsoft.FSharp.Collections.SeqModule.ToArray[T](IEnumerable`1 source) in D:\a\_work\1\s\src\FSharp.Core\seq.fs:line 1027
example-consumer  |    at Microsoft.FSharp.Collections.ArrayModule.OfSeq[T](IEnumerable`1 source) in D:\a\_work\1\s\src\FSharp.Core\array.fs:line 1543
example-consumer  |    at Avro.FSharp.SchemaModule.parse@252-1(Dictionary`2 cache, String ns, JsonElement el)
example-consumer  |    at Avro.FSharp.SchemaModule.parse@252-1(Dictionary`2 cache, String ns, JsonElement el)
example-consumer  |    at Example.KafkaSerde.Message.Deserializer.Confluent.Kafka.IDeserializer<Example.Receipt.T>.Deserialize(ReadOnlySpan`1 data, Boolean isNull, SerializationContext context) in /lib/lib/Data.fs:line 128|
example-consumer  | |CONSUMER|ERROR:Local: Value deserialization error|
example-consumer  | |SCHEMA|EXN:System.Exception: Unknown type: null
example-consumer  |    at Microsoft.FSharp.Core.PrintfModule.PrintFormatToStringThenFail@1448.Invoke(String message)
example-consumer  |    at Avro.FSharp.SchemaModule.parse@316-3.GenerateNext(IEnumerable`1& next)
example-consumer  |    at Microsoft.FSharp.Core.CompilerServices.GeneratedSequenceBase`1.MoveNextImpl() in D:\a\_work\1\s\src\FSharp.Core\seqcore.fs:line 488
example-consumer  |    at Microsoft.FSharp.Core.CompilerServices.GeneratedSequenceBase`1.System.Collections.IEnumerator.MoveNext() in D:\a\_work\1\s\src\FSharp.Core\seqcore.fs:line 530
example-consumer  |    at Microsoft.FSharp.Collections.SeqModule.ToArray[T](IEnumerable`1 source) in D:\a\_work\1\s\src\FSharp.Core\seq.fs:line 1027
example-consumer  |    at Microsoft.FSharp.Collections.ArrayModule.OfSeq[T](IEnumerable`1 source) in D:\a\_work\1\s\src\FSharp.Core\array.fs:line 1543
example-consumer  |    at Avro.FSharp.SchemaModule.parse@252-1(Dictionary`2 cache, String ns, JsonElement el)
example-consumer  |    at Avro.FSharp.SchemaModule.parse@252-1(Dictionary`2 cache, String ns, JsonElement el)
example-consumer  |    at Example.KafkaSerde.Message.Deserializer.Confluent.Kafka.IDeserializer<Example.Receipt.T>.Deserialize(ReadOnlySpan`1 data, Boolean isNull, SerializationContext context) in /lib/lib/Data.fs:line 128|
example-consumer  | |CONSUMER|ERROR:Local: Value deserialization error|
```

Producing data and registering in the registry is fine. Here's what it looks like in the provectus-ui:

```json
{
  "name": "SalesTax",
  "type": [
    "null",
    "double"
  ],
  "default": null
}
```

When I changed that to not an option, everything was fine. 

```
example-consumer  | |CONSUMER|OFFSET:14|MSG:{ Customer = { FirstName = "Jasper"
example-consumer  |                LastName = "Williams"
example-consumer  |                Email = "Jasper_Williams@yahoo.com"
example-consumer  |                PhoneNumber = "0g94akhq74oa"
example-consumer  |                Payment = CreditCard { ExpDate = "Dec/2015"
example-consumer  |                                       Number = "5405-1286-0308-4830"
example-consumer  |                                       CVV = "676" } }
example-consumer  |   OrderID = 3a1ac443-7d31-4e04-938f-8b43d2b88aae
example-consumer  |   OrderNumber = 1501475257561240786L
example-consumer  |   SalesTax = 5M
example-consumer  |   Items =
example-consumer  |    [Soda Large;
example-consumer  |     Combo
example-consumer  |       [DoubleBacon; DoubleBacon; Fries Large; Soda Large; MilkShake Large;
example-consumer  |        Brownie]; Combo [VeganSupreme; Fries Large; Soda Small; Brownie];
example-consumer  |     Soda Large; Soda Large;
example-consumer  |     Combo
example-consumer  |       [DoubleBacon; DoubleBacon; Fries Large; Soda Large; MilkShake Large;
example-consumer  |        Brownie];
example-consumer  |     Combo
example-consumer  |       [Fries Small; Fries Small; Fries Small; MilkShake Small; MilkShake Small;
example-consumer  |        MilkShake Small];
example-consumer  |     Combo
example-consumer  |       [DoubleBacon; DoubleBacon; Fries Large; Soda Large; MilkShake Large;
example-consumer  |        Brownie]]
example-consumer  |   Time = 11/02/2023 14:02:29
example-consumer  |   Total = 322.69M }|FROM:ORDERS [[0]]|
```

What's interesting about this use case is for my generated "fake orders" data I actually always seed the data with `Some 5.00` so you'd think I wouldn't even hit this issue, but for whatever reason, I do. Potentially the internal writerSchema when stringified into JSON is what's throwing that exception.

Thank you again for your work and example code.

**Comment by usix79** [comment]
Thank you for bringing the issue with nullable fields in Avro schemas to our attention. I've identified the bug in the library where null types within union definitions were not being parsed correctly due to missing quotes in the schema's JSON representation.

This has been fixed in the latest release (v1.1.0). Please update your library to this version. Your SalesTax field should now correctly serialize and deserialize as an optional decimal without any further issues.

We appreciate your patience and support in improving our library.

**Comment by wiredsister** [comment]
Thank you @usix79!! I appreciate the patch and your support with this great library.
